src/expr/src/relation/mod.rs

```diff
@@ -1642,6 +1642,24 @@ impl MirRelationExpr {
             .chain(rest.into_iter().flatten())
     }
 
+    /// Iterative pre-order visitor.
+    pub fn visit_pre<'a, F: FnMut(&'a Self)>(&'a self, mut f: F) {
```

Comment:
The main thing I wanted to have access to was a certain-to-be-iterative implementation. If we can fit that in the trait, then I'm delighted, but if we cannot fit it in the trait then I would use the body of this method in-line in its use cases. Iirc the last time I tried to tweak the implementation I ran in to lifetime issues from the trait, though I have to admit I did not try especially hard in this case. If we can put this implementation in the trait implementation, great news!

Reply:
Looking at it, the `VisitChildren` trait does not expose a `children()` iterator, and definitely not a double-ended iterator, which makes descending complicated. I think you can use `visit_children` to put the children on the queue, but they'll be handled in reverse order in that case, unless we `.reverse()` that part of the stack. 

So, we could do that in trait!

Reply:
I think we can extend `VisitChildren<T>` with:

```rust
pub fn children(&self) -> impl DoubleEndedIterator<Item = &T>;
pub fn children_mut(&mut self) -> impl DoubleEndedIterator<Item = &mut T>;
```

and use this to implement some `pre_` and `post_` visit methods that are certain to be iterative.

> I think maybe we could change the lifetime on the trait to correlate the value: `&T` lifetime and the `F: FnMut(&T)` lifetime, but they are otherwise unrelated.

I tried that at some point in the past and got errors elsewhere. I'll open a draft PR to discuss based on a specific proposal.